From 41bfbe8ad835cbd5333f03a6492d8711a8105217 Mon Sep 17 00:00:00 2001
From: wamuir <17497234+wamuir@users.noreply.github.com>
Date: Sun, 2 May 2021 09:58:10 -0500
Subject: [PATCH] simplify generation of go protos

---
 tensorflow/go/genop/generate.sh | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/tensorflow/go/genop/generate.sh b/tensorflow/go/genop/generate.sh
index 4aee23f4d22..7b183244fb3 100644
--- a/tensorflow/go/genop/generate.sh
+++ b/tensorflow/go/genop/generate.sh
@@ -61,12 +61,11 @@ fi
 # Ensure that protoc-gen-go is available in $PATH
 # Since ${PROTOC} will require it.
 export PATH=$PATH:${GOPATH}/bin
-mkdir -p ../vendor
 for FILE in ${TF_DIR}/tensorflow/core/framework/*.proto \
     ${TF_DIR}/tensorflow/core/protobuf/*.proto \
     ${TF_DIR}/tensorflow/stream_executor/*.proto; do
   ${PROTOC} \
     -I ${TF_DIR} \
-    --go_out=../vendor \
+    --go_out=${GOPATH}/src \
     $FILE
 done
-- 
2.34.1

